Add kep for coscheduling base on CRD (#42)
kubernetes-sigs/scheduler-plugins: kep/42-podgroup-coscheduling/README.md

	OccupiedBy string `json:"occupiedBy,omitempty"`

	// The number of actively running pods.
	// +optional
	Scheduled uint32 `json:"scheduled"`

	// The number of actively running pods.
	// +optional
	Running uint32 `json:"running"`

	// The number of pods which reached phase Succeeded.
	// +optional
	Succeeded uint32 `json:"succeeded"`

	// The number of pods which reached phase Failed.
	// +optional
	Failed uint32 `json:"failed"`

	// ScheduleStartTime of the group
	ScheduleStartTime metav1.Time `json:"scheduleStartTime"`
}
```

### Controller

We define a controller to reconcile PodGroup status, and we can query the job status through describing the PodGroup. Onece a pod in a group failed, the Group Status is marked Failed. Controller would also help recover from abnormal cases, e.g. batch scheduling is interrupted due to
cluster upgrade.

### Extension points

#### QueueSort

To make sure a group of pods can be scheduled as soon as possible. We implemented this `extension point`. The main progress is as follows:

1. Sort based on Pod Priority
2. Pod does not belong to any group
3. PodGroup creation time
4. Pod creation time

#### PreFilter

> This extension helps pre-filter pods. It is useful, especially when there are not enough resources in a cluster. The overall flow works as below:

Reviewer: This comment is still outstanding.

`helps ... useful...` are a bit verbose and doesn't add extra info to make the words more understandable. Can you rephrase it?

1. Allow the pods that do not belong to any group.

> 2. If there are no groups scheduling, we check resource, if enough, we allow the pod.

Reviewer: Can you rephrase this? I suppose you want to say:

```suggestion
2. If there are no other pending pod groups - say all other pod groups have already been scheduled, we allow the pod when its resource requirement is satisfied.
```

I'm not quite sure what do you mean by `we check resource`, maybe you mean `MinResource`; if so, please reword it.

Reviewer: This comment is still outstanding.

> 3. If there are groups running, we check if the current pod belong the group having the max progress(num(Pods)/minMember), if it is, we allow it.
> 4. Otherwise, we check if the max finished group can still run when allow this pod. If we can, allow it.
> 5. Otherwise, we check if the pod has higher priority compared with the max finished one. If yes, we reject the pod belongs to the group and allow the current one. 

Reviewer: ```suggestion
3. If there are other pod groups not fully scheduled, we check if the pod group current pod belongs to is the closest to its completion - i.e., we check its completion progress by `len(waiting pods of a pod group) / pod group's minMember`.
    - if it is, this pod is allowed.
    - if it's not, we check if allowing it would still make the top-progressed pod group (or you want to say all progressed pod grouped?) scheduled:
        - if so, this pod is allowed
        - if not, we check if the current pod has higher priority than the top-progressed pod group (I guess for convenience, you would still need the `PriorityClass` defined in PodGroup?). If it does, this pod is allowed and the top-progressed pod group gets rejected.
```

And remove bullet 4 & 5.

Reviewer 2: I don't agree to use `len(waiting pods of a pod group) / pod group's minMember` to check if the pg is the closest to its completion. The minMember and the resource requested in pod are different. Like this：
PG1: min 10 waiting 5 --> 50%
PG2 min 100 waiting 60 --> 60%

Reviewer: @cwdsuzhou does the rewording comment https://github.com/kubernetes-sigs/scheduler-plugins/pull/42/files#r481413128 make sense to you?

Reviewer: Regarding @denkensk 's concern, I'd leave it for @cwdsuzhou to comment. Probably it's not a bad idea to also look at the absolute number to reach completion.

Author: Before pods owned by other groups start scheduling, we would check whether the group with the max completion can continue scheduling when we schedule these pods. According `Sort` func. Current method may be more reasonable. Otherwise, the final scheduling results may conflict with the `Sort`.

Reviewer 2: I think it's a good idea to find the podGroup most likely to be successfully scheduled . But I think the current method is not reasonable. Because the different pods may request different resource and with different scale. And also with different priority.
I think we need to think about the pod.resource.requested、absolute number and the priority to find the the podGroup most likely to be successfully scheduled.
@cwdsuzhou @Huang-Wei  WDYT?

> Any pod rejected to run, their group would be added to a denied list with a ttl.

Reviewer: ```suggestion
For any pod that gets rejected, their pod group would be added to a backoff list and get retried until a TTL is met.
```

Reviewer: This comment is still outstanding.

#### Permit

> 1. When number of pods cannot meet the `minMember` defines in the PodGroup, `Wait` is returned. They will be added to cache with TLL(equal to MaxScheduleTime).

Reviewer: ```suggestion
1. When the number of waiting pods in a PodGroup is less than `minMember` (defined in the PodGroup), the status `Wait` is returned. They will be added to cache with TLL (equal to MaxScheduleTime).
```

Reviewer: This comment is still outstanding.

> 2. When number meet that, we would send a signal to permit the pods waiting. 

Reviewer: ```suggestion
2. When the number is equal or greater than `minMember`, send a signal to permit the waiting pods.
```

> We can define `MaxScheduleTime` for a PodGroup. If anyone of the pods times out, the whole group would be rejected.

Reviewer: FYI: it's now defined in the latest codebase, and will be available in `master` branch until #45 gets reviewed.

Author: I think we should keep this.  Our original version supports from both args and the crd.  If the crd did not set the time, the args defines will work.

scene is as bellow:
A job need 10 pods may time out time after 5s, but a job need 10000 pods would like it to be 5min.

Reviewer 2: I think it is better not to let the user to set the `MaxScheduleTime`. Users will prefer to set the `MaxScheduleTime` Greate than it needed to reserve the resource. This will lead the resource to be reserved for a long time.

> A job need 10 pods may time out time after 5s, but a job need 10000 pods would like it to be 5min.
---
We talked about it before. We can use a factor to compute the `WaitingTime`. The factor can be related to the number of pod.  I add Todo in code before.

Author: But for different cluster,  the factor would also be different. 
e.g. 100 nodes or 2000nodes

Reviewer 2: Yes, the size of cluster is also a factor to be considered with the the number of pod. This can be discussed.
And I am still worried about it. @Huang-Wei @cwdsuzhou 
> I think it is better not to let the user to set the MaxScheduleTime. Users will prefer to set the MaxScheduleTime greater than it needed to reserve the resource. This will lead the resource to be reserved for a long time.

> We can define `MaxScheduleTime` for a PodGroup. If anyone of the pods times out, the whole group would be rejected.

Reviewer: ```suggestion
We can define `MaxScheduleTime` for a PodGroup. If any pod times out, the whole pod group would be rejected.
```

#### PostBind

> This extension is mainly used for helping record the PodGroup Status. When pod binds successfully, we would update the scheduling status of a PodGroup.

Reviewer: ```suggestion
This extension is primarily used to record the PodGroup Status. When a pod is bound successfully, we would update the status of its affiliated PodGroup.
```

Reviewer: This comment is still outstanding.

> We can define `MaxScheduleTime` for a PodGroup. If anyone of the pods times out, the whole group would be rejected.

Reviewer: ```suggestion
We can define `MaxScheduleTime` for a PodGroup. If any pod times out, the whole group would be rejected.
```

## Goals
1. Base on the scheduling framework, implement the co-scheduling feature.
2. Define a CRD name PodGroup to help pod scheduling.
 
## Non-Goals
Sort the job when we submit jobs to a cluster. Currently, we can only do this base on pods.
 
## Use Cases
Batch workloads such as Spark jobs, TensorFlow jobs that have to run altogether.

## Design Details

### PodGroup

We define a CRD name PodGroup to help schedule, its definition is as follows:
```go
// PodGroupSpec represents the template of a pod group.
type PodGroupSpec struct {
	// MinMember defines the minimal number of members/tasks to run the pod group;
	// if there's not enough resources to start all tasks, the scheduler
	// will not start anyone.
	MinMember uint32 `json:"minMember"`

	// MinResources defines the minimal resource of members/tasks to run the pod group;
	// if there's not enough resources to start all tasks, the scheduler
	// will not start anyone.

> 	MinResources *v1.ResourceList `json:"minResources,omitempty"`

Reviewer 2: How about if the podgroup is made up of different types of pods? Like ps and trainer in tensorflow. How to set the MinResources ？
@cwdsuzhou

Author: This is the v1alpha1 version. I could add that if need and extend this CRD later. e.g

type  PodGroup struct {
     ...
     Groups [map]PodGroup 
     ...
}

Reviewer: I think one PodGroup should describe a particular type of workload. In terms of tensorflow application, a good model is to have PS and Worker referencing their corresponding (Sub-)PodGroup, and then there can be a (Parent-)PodGroup to wire those two (Sub-)PodGroup(s).

(not quite sure it would work, need a PoC on this)

Reviewer 3: I thought per https://github.com/kubernetes-sigs/scheduler-plugins/pull/42#discussion_r478137253 `MinResources` was going to be removed?

cluster upgrade.

### Extension points

#### QueueSort

To make sure a group of pods can be scheduled as soon as possible. We implemented this `extension point`. The main progress is as follows:

1. Sort based on Pod Priority
2. Pod does not belong to any group
3. PodGroup creation time
4. Pod creation time

#### PreFilter

This extension pre-filters pods to save scheduling cycles. This is especially helpful when there are not enough resources in a cluster. The overall flow works as below:

1. If the pod doesn't belong to a pod group, allow it; otherwise, go to the following steps.
2. If there are no other pending pod groups - say all other pod groups have already been scheduled, we allow the pod when its resource requirement is satisfied.
3. If there are groups running, we check if the current pod belong the group having the max progress(num(Pods)/minMember), if it is, we allow it.
4. Otherwise, we check if the max finished group can still run when allow this pod. If we can, allow it.

> 5. Otherwise, we check if the pod has higher priority compared with the max finished one. If yes, we reject the pod belongs to the group and allow the current one. 

Reviewer: Please see if the wordings in https://github.com/kubernetes-sigs/scheduler-plugins/pull/42/files#r481413128 makes more sense to you.

Author: Sure,  I missed this last time.

## Design Details

### PodGroup

We define a CRD name PodGroup to help schedule, its definition is as follows:
```go
// PodGroupSpec represents the template of a pod group.
type PodGroupSpec struct {
	// MinMember defines the minimal number of members/tasks to run the pod group;
	// if there's not enough resources to start all tasks, the scheduler
	// will not start anyone.
	MinMember uint32 `json:"minMember"`

	// MinResources defines the minimal resource of members/tasks to run the pod group;
	// if there's not enough resources to start all tasks, the scheduler
	// will not start anyone.
	MinResources *v1.ResourceList `json:"minResources,omitempty"`

	// MaxScheduleTime defines the maximal time of members/tasks to wait before run the pod group;
	MaxScheduleTime *metav1.Duration `json:"maxScheduleTime,omitempty"`
}

// PodGroupStatus represents the current state of a pod group.
type PodGroupStatus struct {
	// Current phase of PodGroup.
	Phase PodGroupPhase `json:"phase"`

    // OccupiedBy marks the workload (e.g., deployment, statefulset) UID that occupy the podgroup.
    // It is empty if not initialized.

> 	OccupiedBy string `json:"occupiedBy,omitempty"`

Reviewer 3: As in, owned, like [`OwnerReference`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#ownerreference-v1-meta)?

Author: Yes

Reviewer 3: I think it would be more intuitive to use the more common terminology?

Reviewer: The canonical `meta.OwnerReference` _may_ be used to exercise the concept of hierarchical PodGroups, i.e., a child PodGroup "ownedBy" the other parent PodGroup, so using the term "OccupiedBy" in `status` can distinguish from that semantics IMO - to represent the semantics of PodGroup <-> Workload mapping.

@cwdsuzhou may comment further.

Author: > The canonical `meta.OwnerReference` _may_ be used to exercise the concept of hierarchical PodGroups, i.e., a child PodGroup "ownedBy" the other parent PodGroup, so using the term "OccupiedBy" in `status` can distinguish from that semantics IMO - to represent the semantics of PodGroup <-> Workload mapping.
> 
> @cwdsuzhou may comment further.

Yep ,this field is used for recording the related object e.g. workload or crd. But not absolutely same as the owner reference.

	// The number of actively running pods.
	// +optional
	Running uint32 `json:"running"`

	// The number of pods which reached phase Succeeded.
	// +optional
	Succeeded uint32 `json:"succeeded"`

	// The number of pods which reached phase Failed.
	// +optional
	Failed uint32 `json:"failed"`

	// ScheduleStartTime of the group
	ScheduleStartTime metav1.Time `json:"scheduleStartTime"`
}
```

### Controller

We define a controller to reconcile PodGroup status, and we can query the job status through describing the PodGroup. Once a pod in a group failed, the Group Status is marked Failed. Controller would also help recover from abnormal cases, e.g. batch scheduling is interrupted due to
cluster upgrade.

### Extension points

#### QueueSort

To make sure a group of pods can be scheduled as soon as possible. We implemented this `extension point`. The main progress is as follows:

1. Sort based on Pod Priority
2. Pod does not belong to any group
3. PodGroup creation time
4. Pod creation time

#### PreFilter

This extension pre-filters pods to save scheduling cycles. This is especially helpful when there are not enough resources in a cluster. The overall flow works as below:

> 1. If the pod doesn't belong to a pod group, allow it; otherwise, go to the following steps.

Reviewer 3: How is this "pod belonging to a pod group" relationship expressed?

Reviewer: - pod to PodGroup: pod gets applied with a PodGroup label or annotation
- PodGroup to pod: either implicitly by dynamic computing, or PodGroup's `OccupiedBy` - which refers to the the group of pods' controller (i.e., Deployment or StatefulSet).

@cwdsuzhou can confirm.

	// will not start anyone.
	MinMember uint32 `json:"minMember"`

	// If specified, indicates the PodGroup's priority. "system-node-critical" and
	// "system-cluster-critical" are two special keywords which indicate the
	// highest priorities with the former being the highest priority. Any other
	// name must be defined by creating a PriorityClass object with that name.
	// If not specified, the PodGroup priority will be default or zero if there is no
	// default.
	// +optional
	PriorityClassName string `json:"priorityClassName,omitempty"`

	// MinResources defines the minimal resource of members/tasks to run the pod group;
	// if there's not enough resources to start all tasks, the scheduler
	// will not start anyone.

> 	MinResources *v1.ResourceList `json:"minResources,omitempty"`

Author: In our original version, we fill this field from podSpec. Keeping this is also ok

> 	MinResources *v1.ResourceList `json:"minResources,omitempty"`

Author: If this defined, we would use this pro pre-fileter. If not, we compute the resource requirements from Pod Spec

> 	MinResources *v1.ResourceList `json:"minResources,omitempty"`

Reviewer 2: I'm confused about how to set the `MinResource`. It's the total minimal resource of the PodGroup or the pod?

> 	MinResources *v1.ResourceList `json:"minResources,omitempty"`

Author: Yes

	Phase PodGroupPhase `json:"phase"`

	// OccupiedBy marks the podgroup occupied by which group.

> 	// Owner reference would be used to filled it, if it is empty, if not initialize.

Reviewer: How about:

```
// OccupiedBy marks the workload (e.g., deployment, statefulset) UID that occupy the podgroup.
// It is empty if not initialized.
```

### Controller

> We define a controller to reconcile PodGroup status, and we can query the job status through describing the PodGroup. Onece a pod in a group failed, the Group Status is marked Failed. Controller would also help recover from abnormal cases, e.g. batch scheduling is interrupted due to

Reviewer: This comment is still outstanding.
